walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,236 @@
+   * the mandatory attendees.
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<String> mandatoryAttendees = request.getAttendees();
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
+    long meetingDurationMinutes = request.getDuration();
+
+    List<TimeRange> mandatoryAttendeesBusyTimeRanges =
+        getAllBusyTimeRanges(events, mandatoryAttendees);
+    List<TimeRange> optionalAttendeesBusyTimeRanges =
+        getAllBusyTimeRanges(events, optionalAttendees);
+
+    sortTimeRangesByStart(mandatoryAttendeesBusyTimeRanges);
+    sortTimeRangesByStart(optionalAttendeesBusyTimeRanges);
+
+    List<TimeRange> mandatoryAttendeesFreeTimeRanges =
+        getFreeTimeRanges(mandatoryAttendeesBusyTimeRanges, meetingDurationMinutes);
+    List<TimeRange> optionalAttendeesFreeTimeRanges =
+        getFreeTimeRanges(optionalAttendeesBusyTimeRanges, meetingDurationMinutes);
+
+    if (mandatoryAttendees.isEmpty()) {
+      return optionalAttendeesFreeTimeRanges;
+    } else if (optionalAttendees.isEmpty()) {
+      return mandatoryAttendeesFreeTimeRanges;
+    } else {
+      return findCommonFreeTimeRanges(
+          mandatoryAttendeesFreeTimeRanges,
+          optionalAttendeesFreeTimeRanges,
+          meetingDurationMinutes);
+    }
+  }
+
+  /**
+   * Constructs all the time ranges that any of the mandatory attendees is busy. These time ranges
+   * cannot be part of the potential meeting time.
+   */
+  private List<TimeRange> getAllBusyTimeRanges(
+      Collection<Event> events, Collection<String> mandatoryAttendees) {
+    List<TimeRange> busyTimes = new ArrayList<TimeRange>();
+
+    for (Event event : events) {
+      Collection<String> eventAttendees = event.getAttendees();
+
+      for (String attendee : eventAttendees) {
+        // If the current event's attendee list contains a mandatory attendee,
+        // adds this event's time range to the set of all busy times,
+        // after checking for overlapping time ranges.
+        if (mandatoryAttendees.contains(attendee)) {
+          TimeRange currEventTimeRange = event.getWhen();
+
+          busyTimes = getUpdatedBusyTimeRanges(currEventTimeRange, busyTimes);
+
+          // If any one of the mandatory attendees is in this event's guest list,
+          // this event's time range needs to be considered busy.
+          // Breaks out of the for-each loop because there is no need to look
+          // at the rest of the event's guest list.
+          break;
+        }
+      }
+    }
+    return busyTimes;
+  }
+
+  /**
+   * Compares a given event's time slot with all the existing busy time slots. Merges the current
+   * event time slot with a busy time slot if they overlap. Adds the current event's time slot to
+   * the list of busy slots in the end.
```

Comment:
nit: The last sentence happens only when the current event's time slot doesn't overlap.

Reply:
Updated.